### PR TITLE
1242 allow custom ransomware extension

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Contributes to #1242

Addresses all requested changes. This allows the user to configure a custom file extension for the agent to use when encrypting files.
- The UI has been updated with a field for customizing the file extension
- The default file extension in the UI is `.m0nk3y`
- Empty file extension is allowed
- The ransomware has been updated to use the custom file extension

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by Running the Monkey locally with relevant config/running Island/...
       as well as by running the new unit tests
* [x] If applicable, add screenshots or log transcripts of the feature working
   Screen Shot 2022-08-18 at 11 37 13 AM
   ```console
   $ ls /tmp/random
   file.txt  some_data.tar
   
   $ python3.7 ./infection_monkey.py m0nk3y -s <server:port>
   ...
   2022-08-18 15:41:46,271 [141255:PayloadThread:DEBUG] ransomware_builder.build_ransomware.26: Ransomware configuration:
   {'encryption': {'directories': {'linux_target_dir': '/tmp/random',
                                   'windows_target_dir': ''},
                   'enabled': True,
                   'file_extension': '.m0nk3y.dory'},
    'other_behaviors': {'readme': True}}
   2022-08-18 15:41:46,272 [141255:PayloadThread:INFO] ransomware.run.43: Running ransomware payload
   2022-08-18 15:41:46,272 [141255:PayloadThread:INFO] ransomware._find_files.53: Collecting files in /tmp/random
   2022-08-18 15:41:46,272 [141255:PayloadThread:INFO] ransomware._encrypt_files.57: Encrypting files in /tmp/random
   2022-08-18 15:41:46,272 [141255:PayloadThread:DEBUG] ransomware._encrypt_files.62: Encrypting /tmp/random/file.txt
   2022-08-18 15:41:46,275 [141255:PayloadThread:DEBUG] ransomware._encrypt_files.62: Encrypting /tmp/random/some_data.tar
   2022-08-18 15:41:46,961 [141255:PayloadThread:INFO] readme_dropper.leave_readme.13: Leaving a ransomware README file at /tmp/random/README.txt
   ...
   
   $ ls /tmp/random
   file.txt.m0nk3y.dory  README.txt  some_data.tar.m0nk3y.dory
   ```

   Another run with an empty file extension:
   ```console
   2022-08-18 18:00:50,178 [151176:PayloadThread:DEBUG] ransomware_builder.build_ransomware.25: Ransomware configuration:
   {'encryption': {'directories': {'linux_target_dir': '/tmp/random',
                                   'windows_target_dir': ''},
                   'enabled': True,
                   'file_extension': ''},
    'other_behaviors': {'readme': True}}
   2022-08-18 18:00:50,179 [151176:PayloadThread:INFO] ransomware.run.43: Running ransomware payload
   2022-08-18 18:00:50,179 [151176:PayloadThread:INFO] ransomware._find_files.53: Collecting files in /tmp/random
   2022-08-18 18:00:50,179 [151176:PayloadThread:INFO] ransomware._encrypt_files.57: Encrypting files in /tmp/random
   2022-08-18 18:00:50,179 [151176:PayloadThread:DEBUG] ransomware._encrypt_files.62: Encrypting /tmp/random/file.txt
   2022-08-18 18:00:50,179 [151176:PayloadThread:DEBUG] ransomware._encrypt_files.62: Encrypting /tmp/random/some_data.tar
   2022-08-18 18:00:50,856 [151176:PayloadThread:INFO] readme_dropper.leave_readme.13: Leaving a ransomware README file at /tmp/random/README.txt
   
   $ ls /tmp/random
   file.txt  README.txt  some_data.tar
   ```

### CODE REVIEW

**Comment by cakekoa** [comment]
I could add more allowable file extension characters. It's currently only accepting letters, numbers, and underscores. Dots are allowed as well (chained file extensions)

File: monkey/tests/unit_tests/infection_monkey/payload/ransomware/test_ransomware.py [hunk 1]
```diff
@@ -41,14 +44,28 @@ def inner(
...
+@pytest.fixture
+def ransomware_options_dict(ransomware_file_extension):
+    options = RANSOMWARE_OPTIONS
+    options["encryption"]["file_extension"] = ransomware_file_extension
+    return options
+
+
+@pytest.fixture
+def encrypt_target_directory():
+    with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/payload/ransomware/test_ransomware.py:67]: There's a pytest fixture for temporary paths that might be better.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/payload/ransomware/ransomware_builder.py [hunk 2]
```diff
@@ -40,15 +40,16 @@ def build_ransomware(options: dict):
...
+        encrypt_bytes=flip_bits, new_file_extension=file_extension, chunk_size=CHUNK_SIZE
     )
 
 
-def _build_file_selector():
+def _build_file_selector(file_extension: str):
     targeted_file_extensions = TARGETED_FILE_EXTENSIONS.copy()
-    targeted_file_extensions.discard(EXTENSION)
+    if file_extension:
+        targeted_file_extensions.discard(EXTENSION)
```
[the remark this answers is not in the document]
Comment (cakekoa) [reply monkey/infection_monkey/payload/ransomware/ransomware_builder.py]: Oh, yes, nice catch!

**Review by ilija-lazoroski — APPROVED** [review]
I think it needs a Changelog entry otherwise good job!

**Comment by mssalvatore** [comment]
> I think it needs a Changelog

+1

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**